cmd/gaia/app/export.go

```diff
@@ -0,0 +1,119 @@
+package app
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	distr "github.com/cosmos/cosmos-sdk/x/distribution"
+	"github.com/cosmos/cosmos-sdk/x/gov"
+	"github.com/cosmos/cosmos-sdk/x/mint"
+	"github.com/cosmos/cosmos-sdk/x/slashing"
+	stake "github.com/cosmos/cosmos-sdk/x/stake"
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+// export the state of gaia for a genesis file
+func (app *GaiaApp) ExportAppStateAndValidators(forZeroHeight bool) (appState json.RawMessage, validators []tmtypes.GenesisValidator, err error) {
+	// as if they could withdraw from the start of the next block
+	ctx := app.NewContext(true, abci.Header{Height: app.LastBlockHeight()})
+
+	// prepare for fresh start at zero height
+	if forZeroHeight {
+
+		/* Handle fee distribution state. */
+
+		// withdraw all delegator & validator rewards
+		app.distrKeeper.IterateValidatorDistInfos(ctx, func(_ int64, valInfo distr.ValidatorDistInfo) (stop bool) {
+			err := app.distrKeeper.WithdrawValidatorRewardsAll(ctx, valInfo.OperatorAddr)
+			if err != nil {
+				panic(err)
+			}
+			return false
+		})
+		app.distrKeeper.IterateDelegationDistInfos(ctx, func(_ int64, distInfo distr.DelegationDistInfo) (stop bool) {
+			app.distrKeeper.WithdrawDelegationRewardsAll(ctx, distInfo.DelegatorAddr)
+			return false
+		})
+
+		// delete all distribution infos
+		app.distrKeeper.RemoveValidatorDistInfos(ctx)
+		app.distrKeeper.RemoveDelegationDistInfos(ctx)
+
+		// assert that the fee pool is empty
+		feePool := app.distrKeeper.GetFeePool(ctx)
+		if !feePool.TotalValAccum.Accum.IsZero() {
+			panic("unexpected leftover validator accum")
+		}
+		bondDenom := app.stakeKeeper.GetParams(ctx).BondDenom
+		if !feePool.ValPool.AmountOf(bondDenom).IsZero() {
+			panic(fmt.Sprintf("unexpected leftover validator pool coins: %v", feePool.ValPool.AmountOf(bondDenom)))
+		}
+
+		// reset fee pool height, save fee pool
+		feePool.TotalValAccum.UpdateHeight = 0
+		app.distrKeeper.SetFeePool(ctx, feePool)
+
+		/* Handle stake state. */
+
+		// iterate through validators by power descending, reset bond height, update bond intra-tx counter
+		store := ctx.KVStore(app.keyStake)
+		iter := sdk.KVStoreReversePrefixIterator(store, stake.ValidatorsByPowerIndexKey)
+		counter := int16(0)
+		for ; iter.Valid(); iter.Next() {
+			addr := sdk.ValAddress(iter.Value())
+			validator, found := app.stakeKeeper.GetValidator(ctx, addr)
+			if !found {
+				panic("expected validator, not found")
+			}
+			validator.BondHeight = 0
+			validator.BondIntraTxCounter = counter
+			// AFAICT we do not need to reset unbonding height since it is not used.
+			app.stakeKeeper.SetValidator(ctx, validator)
+			counter++
+		}
+
+		// AFAICT we do not need to reset bond heights since they are unused.
```

Comment:
(see previous reply)

Reply:
hmm - okay well let's tidy this up, should either:
 - open an issue to remove the bond heights, 
 - reset the bond heights here in this PR, or 
 - remove the bond heights as a part of this PR because they just old

Reply:
Agreed, I'll remove them in this PR.

Reply:
Removed.